javascript/ql/src/Statements/UseOfReturnlessFunction.ql

```diff
@@ -85,11 +85,11 @@ predicate alwaysThrows(Function f) {
 /**
  * Holds if the last statement in the function is flagged by the js/useless-expression query.
  */
-predicate alwaysHasNoEffect(Function f) {
-  exists(ReachableBasicBlock entry, DataFlow::Node noEffect |
-    entry = f.getEntryBB() and
-    hasNoEffect(noEffect.asExpr()) and
-    entry.dominates(noEffect.getBasicBlock())
+predicate lastStatementHasNoEffect(Function f) {
+  exists(DataFlow::Node noEffect |
+    noEffect.getContainer() = f and
+    hasNoEffect(noEffect.asExpr()) and 
+    not exists(noEffect.getASuccessor())
```

Comment:
Is this what you want? I don't think it matters whether `noEffect` has a data-flow successor, it's rather the control-flow successors we are interested in. So you probably want to look for control-flow predecessors of the exit node, or something like that. (This criterion isn't quite correct in the presence of `finally` blocks, but I think it's OK for whitelisting purposes.)